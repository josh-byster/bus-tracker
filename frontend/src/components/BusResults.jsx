import React, { Component } from 'react';
import { Table } from 'reactstrap';
import PropTypes from 'prop-types';
import PullToRefresh from 'pulltorefreshjs';
import { getBuses } from '../util/api';
import BusResultRow from './BusResultRow';
import BusInfoModal from './BusInfoModal';

class BusResults extends Component {
  constructor(props) {
    super(props);
    this.state = {
      departures: [],
      validRequest: null,
      modalInfo: {},
      modalOpen: false
    };
  }

  componentDidMount = () => {
    const { stopInfo } = this.props;
    PullToRefresh.init({
      mainElement: '.tracking-page',
      triggerElement: '.info',
      shouldPullToRefresh: () => {
        return stopInfo.stop_id !== undefined;
      },
      onRefresh: this.getData.bind(this)
    });
  };

  componentDidUpdate(prevProps) {
    const { stopInfo, shouldRefresh } = this.props;
    if (
      prevProps.stopInfo.stop_id !== stopInfo.stop_id ||
      (!prevProps.shouldRefresh && shouldRefresh)
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        departures: [],
        validRequest: null,
        modalInfo: {},
        modalOpen: false
      });
      this.getData();
    }
  }

  getData = async () => {
    const { stopInfo, resultCallback } = this.props;
    const { status, departures } = await getBuses(stopInfo.stop_id);
    if (status.code === 200) {
      this.setState({ validRequest: true });
      this.setState({ departures });
    } else {
      this.setState({ validRequest: false });
    }
    resultCallback();
  };

  toggleModal = info => {
    this.setState(state => {
      if (info !== undefined) {
        return { modalOpen: !state.modalOpen, modalInfo: info };
      }
      // Only toggle, don't update info if there is no new info
      return { modalOpen: !state.modalOpen };
    });
  };

  render() {
    const { validRequest, departures, modalInfo, modalOpen } = this.state;
    const { style, stopInfo } = this.props;
    if (validRequest === null) {
      // On first time don't swap out elements unnecessarily. Just render div.
      return <div />;
    }
    if (validRequest === false) {
      return <h4>This page cannot be loaded.</h4>;
    }

    if (validRequest === true && departures.length === 0) {
      return <h4 className="no-bus">No buses coming in the next hour.</h4>;
    }
    return (
      <div style={style}>
        <BusInfoModal
          busInfo={modalInfo}
          isOpen={modalOpen}
          toggle={this.toggleModal}
          stopInfo={stopInfo}
        />
        <Table>
          <thead>
            <tr>
              <th id="bus-name">Bus Name</th>
              <th id="mins-left">Mins Left</th>
              <th id="eta">ETA</th>
              <th id="last-location">Last Location</th>
            </tr>
          </thead>
          <tbody>
            {departures.map((element, key) => {
              return (
                <BusResultRow
                  info={element}
                  toggleModal={this.toggleModal}
                  key={key}
                  elementOrder={key}
                />
              );
            })}
          </tbody>
        </Table>
      </div>
    );
  }
}

BusResults.propTypes = {
  stopInfo: PropTypes.object.isRequired,
  resultCallback: PropTypes.func.isRequired,
  style: PropTypes.object.isRequired,
  shouldRefresh: PropTypes.bool.isRequired
};
export default BusResults;
