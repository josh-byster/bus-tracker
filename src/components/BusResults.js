import React, { Component } from "react";
import { getBuses } from "../util/api";
import BusResultRow from "./BusResultRow";
import { Table } from "reactstrap";
import PropTypes from "prop-types";
import BusInfoModal from "../components/BusInfoModal";
class BusResults extends Component {
  constructor(props) {
    super(props);
    this.state = {
      departures: [],
      validRequest: null,
      modalInfo: {},
      modalOpen: false
    };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.stopInfo.stop_id !== this.props.stopInfo.stop_id) {
      this.setState({
        departures: [],
        validRequest: null,
        modalInfo: {},
        modalOpen: false
      });
      this.getData(this.props.stopInfo.stop_id);
    }
  }
  getData = async stop_id => {
    const { status, departures } = await getBuses(stop_id);
    if (status.code === 200) {
      this.setState({ validRequest: true });
      this.setState({ departures: departures });
    } else {
      this.setState({ validRequest: false });
    }
    this.props.resultCallback();
  };

  toggleModal = info => {
    this.setState(state => {
      if (info !== undefined) {
        return { modalOpen: !state.modalOpen, modalInfo: info };
      }
      // Only toggle, don't update info if there is no new info
      return { modalOpen: !state.modalOpen };
    });
  };

  render() {
    if (this.state.validRequest === null) {
      // On first time don't swap out elements unnecessarily. Just render div.
      return <div />;
    }
    if (this.state.validRequest === false) {
      return <h4>This page cannot be loaded.</h4>;
    }

    if (
      this.state.validRequest === true &&
      this.state.departures.length === 0
    ) {
      return <h4 className="no-bus">No buses coming in the next hour.</h4>;
    }
    return (
      <div style={this.props.style}>
        <BusInfoModal
          busInfo={this.state.modalInfo}
          isOpen={this.state.modalOpen}
          toggle={this.toggleModal}
          stopInfo={this.props.stopInfo}
        />
        <Table>
          <thead>
            <tr>
              <th id="bus-name">Bus Name</th>
              <th id="mins-left">Mins Left</th>
              <th id="eta">ETA</th>
              <th id="last-location">Last Location</th>
            </tr>
          </thead>
          <tbody>
            {this.state.departures.map((element, key) => {
              return (
                <BusResultRow
                  info={element}
                  toggleModal={this.toggleModal}
                  key={key}
                  elementOrder={key}
                />
              );
            })}
          </tbody>
        </Table>
      </div>
    );
  }
}

BusResults.propTypes = {
  stopInfo: PropTypes.object.isRequired,
  resultCallback: PropTypes.func.isRequired,
  style: PropTypes.object
};
export default BusResults;
