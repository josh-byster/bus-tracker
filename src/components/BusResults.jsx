import React, { Component } from 'react';
import { Table } from 'reactstrap';
import PropTypes from 'prop-types';
import PullToRefresh from 'pulltorefreshjs';
import { getBuses } from '../util/api';
import BusResultRow from './BusResultRow';
import BusInfoModal from './BusInfoModal';

class BusResults extends Component {
  constructor(props) {
    super(props);
    this.state = {
      departures: [],
      validRequest: null,
      modalInfo: {},
      modalOpen: false
    };
  }

  componentDidMount = () => {
    const { stopInfo } = this.props;
    PullToRefresh.init({
      mainElement: '.tracking-page',
      triggerElement: '.info',
      shouldPullToRefresh: () => {
        return stopInfo.stop_id !== undefined;
      },
      onRefresh: this.getData.bind(this)
    });
  };

  componentDidUpdate(prevProps) {
    const { stopInfo } = this.props;
    if (prevProps.stopInfo.stop_id !== stopInfo.stop_id) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        departures: [],
        validRequest: null,
        modalInfo: {},
        modalOpen: false
      });
      this.getData();
    }
  }

  getData = async () => {
    const { stopInfo, resultCallback } = this.props;
    const { status, departures } = await getBuses(stopInfo.stop_id);
    if (status.code === 200) {
      this.setState({ validRequest: true });
      this.setState({ departures });
    } else {
      this.setState({ validRequest: false });
    }
    resultCallback();
  };

  toggleModal = info => {
    this.setState(state => {
      if (info !== undefined) {
        return { modalOpen: !state.modalOpen, modalInfo: info };
      }
      // Only toggle, don't update info if there is no new info
      return { modalOpen: !state.modalOpen };
    });
  };

  render() {
    if (this.state.validRequest === null) {
      // On first time don't swap out elements unnecessarily. Just render div.
      return <div />;
    }
    if (this.state.validRequest === false) {
      return <h4>This page cannot be loaded.</h4>;
    }

    if (
      this.state.validRequest === true &&
      this.state.departures.length === 0
    ) {
      return <h4 className="no-bus">No buses coming in the next hour.</h4>;
    }
    return (
      <div style={this.props.style}>
        <BusInfoModal
          busInfo={this.state.modalInfo}
          isOpen={this.state.modalOpen}
          toggle={this.toggleModal}
          stopInfo={this.props.stopInfo}
        />
        <Table>
          <thead>
            <tr>
              <th id="bus-name">Bus Name</th>
              <th id="mins-left">Mins Left</th>
              <th id="eta">ETA</th>
              <th id="last-location">Last Location</th>
            </tr>
          </thead>
          <tbody>
            {this.state.departures.map((element, key) => {
              return (
                <BusResultRow
                  info={element}
                  toggleModal={this.toggleModal}
                  key={key}
                  elementOrder={key}
                />
              );
            })}
          </tbody>
        </Table>
      </div>
    );
  }
}

BusResults.propTypes = {
  stopInfo: PropTypes.object.isRequired,
  resultCallback: PropTypes.func.isRequired,
  style: PropTypes.object
};
export default BusResults;
