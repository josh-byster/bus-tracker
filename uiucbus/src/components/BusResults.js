import React, { Component } from "react";
import { getBuses } from "../util/api";
import BusResultRow from "./BusResultRow";
import { Table } from "reactstrap";
class BusResults extends Component {
  constructor(props) {
    super(props);
    this.state = {
      departures: []
    };
    this.getData(props.stop_id);
  }
  getData = async stop_id => {
    const { status, rqst, departures } = await getBuses(stop_id);
    this.setState({ departures: departures });
  };
  render() {
    return (
      <div>
        <Table>
          <thead>
            <tr>
              <th>Bus Name</th>
              <th>Mins Left</th>
              <th>ETA</th>
              <th>Last Location</th>
            </tr>
          </thead>
          <tbody>
            {this.state.departures.map((element, key) => {
              return <BusResultRow info={element} key={key} />;
            })}
          </tbody>
        </Table>
      </div>
    );
  }
}

export default BusResults;
