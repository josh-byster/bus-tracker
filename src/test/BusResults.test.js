import React from "react";
import expect from "expect";
import { shallow } from "enzyme";
import BusResults from "../components/BusResults";

// setup file
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });
var stopInfo = {
  stop_id: "IU"
};
describe("Bus Result Component:", () => {
  it("renders without crash", () => {
    expect(shallow(<BusResults stopInfo={stopInfo} />).length).toEqual(1);
  });

  it("starts with a modal closed", () => {
    expect(
      shallow(<BusResults stopInfo={stopInfo} />)
        .setProps({ stopInfo: { stop_id: "PLAZA" } })
        .state().modalOpen
    ).toBeFalsy();
  });
});
