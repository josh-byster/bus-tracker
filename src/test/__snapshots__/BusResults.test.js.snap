// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bus Result Component: renders many buses with mock data  1`] = `
<div>
  <div />
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          id="bus-name"
        >
          Bus Name
        </th>
        <th
          id="mins-left"
        >
          Mins Left
        </th>
        <th
          id="eta"
        >
          ETA
        </th>
        <th
          id="last-location"
        >
          Last Location
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#fcee1f",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            100N Yellow
          </b>
        </td>
        <td>
          5m
        </td>
        <td
          className="no-wrap"
        >
          5:32:40
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#cccccc",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            130S Silver Limited
          </b>
        </td>
        <td>
          7m
        </td>
        <td
          className="no-wrap"
        >
          5:34:15
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#cccccc",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            130N Silver Limited
          </b>
        </td>
        <td>
          9m
        </td>
        <td
          className="no-wrap"
        >
          5:36:15
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#5a1d5a",
            "color": "#ffffff",
          }
        }
      >
        <td>
          <b>
            220S Illini Limited
          </b>
        </td>
        <td>
          18m
        </td>
        <td
          className="no-wrap"
        >
          5:45:00
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#fcee1f",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            100S Yellow
          </b>
        </td>
        <td>
          19m
        </td>
        <td
          className="no-wrap"
        >
          5:46:40
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#5a1d5a",
            "color": "#ffffff",
          }
        }
      >
        <td>
          <b>
            220N Illini Limited
          </b>
        </td>
        <td>
          20m
        </td>
        <td
          className="no-wrap"
        >
          5:47:00
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#cccccc",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            130S Silver Limited
          </b>
        </td>
        <td>
          26m
        </td>
        <td
          className="no-wrap"
        >
          5:53:00
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
      <tr
        className="resultRow"
        style={
          Object {
            "backgroundColor": "#cccccc",
            "color": "#000000",
          }
        }
      >
        <td>
          <b>
            130N Silver Limited
          </b>
        </td>
        <td>
          28m
        </td>
        <td
          className="no-wrap"
        >
          5:55:00
        </td>
        <td>
          <button
            aria-label={null}
            className="btn btn-success"
            onClick={[Function]}
            type="button"
          >
            Location
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Bus Result Component: renders no buses coming 1`] = `
<h4>
  No buses coming in the next hour.
</h4>
`;

exports[`Bus Result Component: renders the proper h4 with invalid request 1`] = `
<h4>
  This page cannot be loaded.
</h4>
`;
